### PR TITLE
[GraphQL][DotMove][3/n] Adds `External` resolver

### PR DESCRIPTION
**Pull request by manolisliolios**
## Description 

Introduces the external resolution, which requests "names" from a different graphql node. 

In reality, that'll be a mainnet RPC endpoint being used. 

## Test plan 

e2e tests updated to start a secondary service that utilizes external resolution, and depend on that:

```
cargo nextest run --package sui-graphql-rpc --test dot_move_e2e --features pg_integration
```

## Stack

- #18797 
- #18774 
- #18770 

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-types/src/collection_types.rs
Comment (amnn) [inline crates/sui-types/src/collection_types.rs:1]: (This is also something that you could split out into its own PR ;) -- that way it would be more obvious that we're missing tests for it!)

Comment (amnn) [reply crates/sui-types/src/collection_types.rs:1]: No need to split it out now, but tests would seem to be useful.

File: crates/sui-types/src/collection_types.rs [hunk 1]
```diff
@@ -12,6 +12,18 @@ pub struct VecMap<K, V> {
...
 }
 
+impl<K: PartialEq, V> VecMap<K, V> {
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.contents.iter().find_map(|entry| {
+            if &entry.key == key {
+                Some(&entry.value)
+            } else {
+                None
+            }
```
Comment (amnn) [inline crates/sui-types/src/collection_types.rs:22]:
```suggestion
            (&entry.key == key).then_some(&entry.value)
```

File: crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    type Error = Error;
+
+    /// This function queries the mainnet API to fetch the app records for the requested names.
+    /// This is part of the data loader, so all queries are bulked-up to the maximum of {config.page_limit}.
+    /// We handle the cases where individual queries fail, to ensure that a failed query cannot affect
+    /// a successful one.
+    async fn load(&self, keys: &[Name]) -> Result<HashMap<Name, AppRecord>, Error> {
+        let Some(mainnet_api_url) = self.config.mainnet_api_url.as_ref() else {
+            return Err(Error::DotMove(
+                DotMoveServiceError::MainnetApiUrlUnavailable,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs:85]: Ditto -- `External` vs `Mainnet`.

**Review by amnn — COMMENTED** [review]
Very nice -- main comment here is related to distancing ourselves from the idea that external resolution always talks to mainnet.

File: crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+use crate::{error::Error, types::base64::Base64};
+
+use super::config::{AppRecord, DotMoveConfig, DotMoveServiceError, Name};
+
+/// GraphQL fragment to query the values of the dynamic fields.
+const QUERY_FRAGMENT: &str =
+    "fragment RECORD_VALUES on DynamicField { value { ... on MoveValue { bcs } } }";
+
+pub(crate) struct MainnetNamesLoader {
```
[the remark this answers is not in the document]
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs:17]: Completely agreed... lots of bias from the early days of the design!

File: crates/sui-types/src/collection_types.rs (hunk 1 above)
Comment (manolisliolios) [reply crates/sui-types/src/collection_types.rs:22]: You can't get bored with Rust. Whenever I think I've exhausted all the "builder-style" calls, there are more! 😆

File: crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            mapping.insert(name.clone(), index);
+
+            let field_str = format!(
+                r#"{}: dynamicField(name: {{ type: "{}::name::Name", bcs: {} }}) {{ ...RECORD_VALUES }}"#,
+                fetch_key(&index),
+                self.config.package_address,
+                bcs_base64
+            );
+
+            result.push_str(&field_str);
```
[the remark this answers is not in the document]
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/types/dot_move/data_loader.rs:55]: We're writing to a String, not a buffer, so the `write` macro doesn't seem to do what we want.
assignment is not needed anyways, I just did an immediate `result.push_str(&format!()` to avoid the assignment instead.

File: crates/sui-types/src/collection_types.rs
Comment (manolisliolios) [reply crates/sui-types/src/collection_types.rs:1]: Added a basic test.

File: crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+            // retain the mapping here (id to bcs representation, so we can pick the right response later on)
+            mapping.insert(name.clone(), index);
+
+            result.push_str(&format!(
+                r#"{}: dynamicField(name: {{ type: "{}::name::Name", bcs: {} }}) {{ ...RECORD_VALUES }}"#,
+                fetch_key(&index),
+                self.config.package_address,
+                bcs_base64
+            ));
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs:60]: Here is a playground link demonstrating what I mentioned about using `write!` directly before:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66009e35c7c6d3e53173a813d5531ebd

The key is that you need to have `std::fmt::Write` in scope, and you are writing to the mutable borrow of the string.

File: crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                continue;
+            };
+
+            let Some(app_record) = bcs::from_bytes::<AppRecord>(&bytes.0).ok() else {
+                continue;
+            };
+
+            // only insert the record if it is a valid `app_record`
+            results.insert(k.clone(), app_record);
+        }
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs:149]: Here's a more compact way to do this:

```suggestion
        let results = HashMap::from_iter(
            mapping
                .into_iter()
                .filter_map(|k, idx| {
                    let bcs = names.get(&fetch_key(idx))??;
                    let Base64(bytes) = Base64::from_str(&bcs.value.bcs).ok()?;
                    let app_record: AppRecord = bcs::from_bytes(&bytes).ok()?;
                    Some((k, app_record))
                })
        );
```

Although I wonder whether if some of these steps fail, you should be producing an error instead, rather than dropping the value? Not sure -- that only works if the error can't be triggered by user input (i.e. it's an internal error of some kind.

File: crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+        Ok(results)
+    }
+}
+
+impl Default for ExternalNamesLoader {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs:155]: Out of interest, what do you use this default impl for?

File: crates/sui-graphql-rpc/tests/dot_move_e2e.rs [hunk 8]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-#[cfg(feature = "pg_integration")]
+// #[cfg(feature = "pg_integration")]
```
Comment (amnn) [inline crates/sui-graphql-rpc/tests/dot_move_e2e.rs:4]: nit -- add back

File: crates/sui-graphql-rpc/src/server/builder.rs [hunk 9]
```diff
@@ -427,7 +428,8 @@ impl ServerBuilder {
             .context_data(zklogin_config)
             .context_data(metrics.clone())
             .context_data(config.clone())
-            .context_data(move_registry_config.clone());
+            .context_data(move_registry_config.clone())
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/server/builder.rs:431]: This is not something that we need to address in this PR, but as a future clean-up we need to standardise how we deal with configs in context data. Today we add the entire `ServiceConfig` to the context data, and then we separately add some of its constituent parts (`zklogin_config`, `name_service_config`, and now `move_registry_config`) -- not clear which way is better (splitting up `ServiceConfig` or not), but having both methods of accessing these configs is going to get confusing.

File: crates/sui-graphql-rpc/src/types/dot_move/named_move_package.rs [hunk 10]
```diff
@@ -26,54 +33,96 @@ impl NamedMovePackage {
...
+) -> Result<Option<MovePackage>, Error> {
+    if config.external_api_url.is_none() {
+        return Err(MoveRegistryError::ExternalApiUrlUnavailable.into());
     }
+
+    let identifier: ChainIdentifier = *ctx.data_unchecked();
+
+    let Some(chain_id) = identifier.0 else {
+        return Err(MoveRegistryError::ChainIdentifierUnavailable.into());
+    };
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dot_move/named_move_package.rs:86]: nit: Combine this into one statement
```suggestion
    let ChainIdentifier(Some(&chain_id)) = ctx.data_unchecked() else {
        return Err(MoveRegistryError::ChainIdentifierUnavailable.into());
    };
```

File: crates/sui-graphql-rpc/src/types/dot_move/named_move_package.rs [hunk 11]
```diff
@@ -26,54 +33,96 @@ impl NamedMovePackage {
...
+
+    let MoveRegistryDataLoader(loader) = &ctx.data_unchecked();
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dot_move/named_move_package.rs:88]: nit: is the borrow necessary here? `data_unchecked` already returns a reference.

**Review by amnn — APPROVED** [review]
This is good to go modulo nits, thanks @manolisliolios !

**Comment by amnn** [comment]
(But do take a look at the thing about using `write!` instead of `.push_str(&format!(...))`.

File: crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs (hunk 7 above)
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs:155]: Oh, it's not needed. I can't remember why I added it at some point. Removed it.

File: crates/sui-graphql-rpc/src/types/dot_move/named_move_package.rs (hunk 11 above)
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/types/dot_move/named_move_package.rs:88]: Nice catch, it isn't! Removed it.

File: crates/sui-graphql-rpc/src/server/builder.rs (hunk 9 above)
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/server/builder.rs:431]: Fair argument and we should indeed figure out a more reusable way of handling these.

File: crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs (hunk 5 above)
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs:60]: Ahh you're right... That's what it was complaining for (missing `std::fmt::Write`). Sorry for having you say it to me one million times 🫠

File: crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs (hunk 6 above)
Comment (manolisliolios) [reply crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs:149]: Thanks, love the more compact approach.

We'll error out when we try to populate things in the "controllers". Since this data loader is doing a bulk operation here (so it's likely that it is serving multiple requests), I didn't want irrelevant requests to error out that's why we ignore singular name errors on this layer.

It's a different story for full request errors, there we indeed do error out.
